src/Npgsql/TaskTimeoutAndCancellation.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Util;
+
+namespace Npgsql;
+
+static class TaskTimeoutAndCancellation
+{
+    /// <summary>
+    /// Utility that executes a non-cancellable task while allowing to timeout and/or cancel awaiting for it.
+    /// If the given task does not complete within <paramref name="timeout"/>, a <see cref="TimeoutException"/> is thrown.
+    /// </summary>
+    /// <param name="getTaskFunc">Gets a <see cref="Task{TResult}"/>.</param>
+    /// <param name="timeout">The timeout after which the <see cref="Task"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+    /// <typeparam name="TResult">The result <see cref="Type"/>.</typeparam>
+    /// <returns>The <see cref="Task{TResult}"/> representing the asynchronous wait.</returns>
+    internal static async Task<TResult> WaitAsync<TResult>(Func<CancellationToken, Task<TResult>> getTaskFunc, NpgsqlTimeout timeout, CancellationToken cancellationToken)
+    {
+        Task<TResult>? task = default;
+        await WaitAsync(ct => (Task)(task = getTaskFunc(ct)), timeout, cancellationToken);
+        return await task!;
+    }
+
+    /// <summary>
+    /// Utility that executes a non-cancellable task while allowing to timeout and/or cancel awaiting for it.
+    /// If the given task does not complete within <paramref name="timeout"/>, a <see cref="TimeoutException"/> is thrown.
+    /// </summary>
+    /// <param name="getTaskFunc">Gets a <see cref="Task"/>.</param>
+    /// <param name="timeout">The timeout after which the <see cref="Task"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+    /// <returns>The <see cref="Task"/> representing the asynchronous wait.</returns>
+    internal static async Task WaitAsync(Func<CancellationToken, Task> getTaskFunc, NpgsqlTimeout timeout,
```

Comment:
It seems like the naming (and documentation) should express the fact that this method swallows unobserved exceptions, especially given the .NET 6.0 WaitAsync which has the same name but does not... Maybe WaitAsyncAndPreventUnobservedExceptions?

Reply:
If I understand correctly, a method's name that API accepts `Func<CancellationToken, Task> getTaskFunc` should start with "Execute". Then if I rename to `ExecuteWithTimeoutAndFakeCancellationWhilePreventingUnobservedExceptions` and update the comments, would that satisfy such a requirement?

Reply:
I'm not 100% hung up on specific naming - just trying to make the code clearer. I think it would be better to separate out the concern of applying a timeout via the cancellation token (previously called ExecuteWithTimeout) and the concern of preventing unobserved exceptions. And yeah, Execute seems like better naming than Wait since .NET's WaitAsync accepts an already running Task, whereas Execute accepts a lambda which will be executed and return that Task.

ExecuteWithTimeoutAndFakeCancellationWhilePreventingUnobservedExceptions is quite a mouthful :) We could simply leave two separate methods (ExecuteWithTimeout, WaitAsyncWhilePreventingUnobservedExceptions) and call one over the other in the two call sites we have. Or if you feel strongly that there should be one method which takes care of both (while ideally calling a local ExecuteWithTimeout method), that's OK for me too.

Reply:
We should not call one over another because
- doing so in 2 places would be repeating yourself,
- splitting ExecuteWithTimeout from PreventingUnobservedExceptions would promote future misuse and more unobserved exceptions.

Considering that the functionality is private to NpgsqlConnector I'd vote for a shorter name like `ExecuteWithTimeoutAndCancellation` leaving the rest to the comment.

For a general-purpose API we'd add a parameter "preventUnobservedExceptions" that would self-document the API but we should not do that without a use case.

Reply:
> splitting ExecuteWithTimeout from PreventingUnobservedExceptions would promote future misuse and more unobserved exceptions.

Technically I don't think that's true - ExecuteWithTimout simply implements timeout via the cancellation token, and by itself shouldn't create any misuse.

But I think we've spent too much time discussing this. Proper commenting should help make things clear.

Reply:
Respecting your opinion on how the code has to be structured: let me take another shot at this.

Reply:
The problem is that we need to "unobserve" the Task produced by `getTaskFunc`. That makes for an awkward `ExecuteWithTimeout` interface to satisfy the need to await on the composite Task and at the same time returning `getTaskFunc` Task for later optional chaining into `PreventingUnobservedExceptions`.

Feel free to make your own update if I am missing the right way to do this.

Reply:
I'm OK with the code as it is... Let's get it merged.